### PR TITLE
Closes #1759: Copy telemetry events to Firebase

### LINKED ISSUE #1759: Firebase Telemetry to include some Rocket specific events

### PR DESCRIPTION
**Pull request by cnevinc**
Closes #1759

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java [hunk 1]
```diff
@@ -284,8 +286,20 @@ public static void settingsEvent(String key, String value) {
...
     }
 
-    public static void settingsClickEvent(String key) {
+    public static void settingsClickEvent(Context context, String key) {
         TelemetryEvent.create(Category.ACTION, Method.CLICK, Object.SETTING, key).queue();
+
+        if (context == null) {
+            return;
+        }
+        String pref_key_storage_save_downloads_to = context.getString(R.string.pref_key_storage_save_downloads_to);
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: Just curious: If we're only interested in these events, why not also apply this key check for Telemetry?

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java [hunk 2]
```diff
@@ -294,11 +308,22 @@ public static void settingsLearnMoreClickEvent(String source) {
...
         TelemetryEvent.create(Category.ACTION, Method.CHANGE, Object.SETTING, key)
                 .extra(Extra.TO, value)
                 .extra(Extra.DEFAULT, Boolean.toString(isDefault))
                 .queue();
+
+        if (context == null) {
+            return;
+        }
+        String pref_key_turbo_mode = context.getString(R.string.pref_key_turbo_mode);
+        if (key.equals(pref_key_turbo_mode)) {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: I think key is always R.string.pref_key_locale for this function.

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java [hunk 3]
```diff
@@ -398,10 +423,16 @@ public static void showMenuHome() {
 
     public static void showTabTrayHome() {
         TelemetryEvent.create(Category.ACTION, Method.SHOW, Object.TABTRAY, Value.HOME).queue();
+
+        final Bundle params = new Bundle();
+        params.putString(FirebaseHelper.EventParam.VALUE, Value.HOME);
+        FirebaseHelper.event(Method.SHOW + Object.MENU, Value.HOME);
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: , params);?

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java [hunk 4]
```diff
@@ -430,22 +463,34 @@ public static void showPanelHistory() {
...
     }
 
     public static void menuTurboChangeTo(boolean enable) {
         TelemetryEvent.create(Category.ACTION, Method.CHANGE, Object.MENU, Value.TURBO)
                 .extra(Extra.TO, Boolean.toString(enable))
                 .queue();
+
+        final Bundle extra = new Bundle();
+        extra.putString(Extra.TO, Boolean.toString(enable));
+        FirebaseHelper.event(Method.CHANGE + Object.MENU, Value.CAPTURE, extra);
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: Value.CAPTURE should be Value.TURBO

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java [hunk 5]
```diff
@@ -466,22 +511,35 @@ public static void clickToolbarReload() {
...
     public static void clickTopSiteOn(int index) {
         TelemetryEvent.create(Category.ACTION, Method.OPEN, Object.HOME, Value.LINK)
                 .extra(Extra.ON, Integer.toString(index))
                 .queue();
+
+        final Bundle extra = new Bundle();
+        extra.putString(Extra.ON, Integer.toString(index));
+        FirebaseHelper.event(Method.OPEN + Object.HOME, Value.LINK, extra);
+
         TelemetryEvent.create(Category.ACTION, Method.ADD, Object.TAB, Value.TOPSITE)
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: Are you also copying this too?

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java [hunk 6]
```diff
@@ -613,75 +689,121 @@ public static void openCapture() {
...
     }
 
     public static void clickRateAppNotification(String value) {
         TelemetryEvent.create(Category.ACTION, Method.CLICK, Object.FEEDBACK, value)
                 .extra(Extra.SOURCE, TelemetryWrapper.Extra_Value.NOTIFICATION)
                 .queue();
+
+        final Bundle extra = new Bundle();
+        extra.putString(Extra.SOURCE, Extra_Value.NOTIFICATION);
+        FirebaseHelper.event(Method.CLICK + Object.PROMOTE_SHARE, value, extra);
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: Object.FEEDBACK here

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java [hunk 7]
```diff
@@ -613,75 +689,121 @@ public static void openCapture() {
...
     }
 
     public static void clickRateAppNotification() {
         TelemetryEvent.create(Category.ACTION, Method.CLICK, Object.FEEDBACK)
                 .extra(Extra.SOURCE, TelemetryWrapper.Extra_Value.NOTIFICATION)
                 .queue();
+
+        final Bundle extra = new Bundle();
+        extra.putString(Extra.SOURCE, Extra_Value.NOTIFICATION);
+        FirebaseHelper.event(Method.CLICK + Object.PROMOTE_SHARE, null, extra);
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: Object.FEEDBACK here

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java [hunk 8]
```diff
@@ -613,75 +689,121 @@ public static void openCapture() {
...
+
+        final Bundle extra = new Bundle();
+        extra.putString(Extra.SOURCE, source);
+        FirebaseHelper.event(Method.CLICK + Object.PROMOTE_SHARE, value, extra);
     }
 
     public static void showPromoteShareDialog() {
         TelemetryEvent.create(Category.ACTION, Method.SHOW, Object.PROMOTE_SHARE).queue();
+
+        FirebaseHelper.event(Method.CLICK + Object.PROMOTE_SHARE, null);
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: Method.SHOW here

File: app/src/main/java/org/mozilla/focus/utils/FirebaseHelper.java [hunk 9]
```diff
@@ -202,5 +208,15 @@ protected Void doInBackground(Void... voids) {
         return map;
     }
 
+    public static void event(String key, String value) {
+        event(key, value, null);
+    }
+
+    public static void event(String key, String value, Bundle extra) {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/utils/FirebaseHelper.java]:
```
Maybe event(String method, String object, String value, Bundle extra) {
   final String key = method + object 
}
```
will make your life easier.
In the meanwhile, maybe we need a `-` or `_` to avoid duplicate method + object possibilities? (Quite unlikely though.)

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java (hunk 1 above)
Comment (cnevinc) [reply app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: In the document, only a subset of events is sent to Firebase. So I exclude others here.

File: app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+    // limitation for event:
+    // Event names can be up to 40 characters long, may only contain alphanumeric characters and
+    // underscores ("_"), and must start with an alphabetic character. The "firebase_", "google_"
+    // and "ga_" prefixes are reserved and should not be used.
+    // see: https://firebase.google.com/docs/reference/android/com/google/firebase/analytics/FirebaseAnalytics.Event
+    static final int MAX_LENGTH_EVENT_NAME = 40;
+    static final String EVENT_NAME_SEPARATOR = "__";
+
+    // limitation fo param
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java]: NITNIT: `limitation fo param:`

File: app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    private String eventName;
+    private Bundle eventParam;
+
+
+    private FirebaseEvent(@NonNull String category, @NonNull String method, @Nullable String object, @Nullable String value) {
+
+        // append null string will append 'null'
+        // Android Studio also said we don't need to use a String builder here.
+        this.eventName = method + EVENT_NAME_SEPARATOR + object + EVENT_NAME_SEPARATOR + value;
+        if (!AppConstants.isReleaseBuild() && this.eventName.length() > MAX_LENGTH_EVENT_NAME) {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java]: What happens when we're in production, and the length is too long? The Event is still set and will be queued. (Will Firebase trim it? Or some exception or ignored? Which is our expected behavior?)

File: app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        this.eventParam.putString(safeParamLength(name, MAX_LENGTH_PARAM_NAME),
+                safeParamLength(value, MAX_LENGTH_PARAM_VALUE));
+
+        return this;
+    }
+
+
+    private static String safeParamLength(@NonNull final String str, final int end) {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java]: NIT: variable name end seems weird to me.

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java [hunk 13]
```diff
@@ -464,24 +511,38 @@ public static void clickToolbarReload() {
...
         TelemetryEvent.create(Category.ACTION, Method.PIN_SHORTCUT, Object.TOOLBAR, Value.LINK).queue();
+
+        FirebaseEvent.create(Category.ACTION, Method.PIN_SHORTCUT, Object.TOOLBAR, Value.LINK).queue(context);
     }
 
-    public static void clickToolbarCapture() {
+    public static void clickToolbarCapture(Context context) {
         TelemetryEvent.create(Category.ACTION, Method.CLICK, Object.TOOLBAR, Value.CAPTURE)
-                .extra(Extra.VERSION, Integer.toString(2))
+                .extra(Extra.VERSION, Integer.toString(TELEMETRY_VERSION))
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: I'm not sure about this. This is about toolbarCapture V2, but not telemetry V2. (All telemetry has its own version, in this case, the position of toolbarcapture have changed in 2.0)

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java [hunk 14]
```diff
@@ -464,24 +511,38 @@ public static void clickToolbarReload() {
...
                 .queue();
+
+        FirebaseEvent.create(Category.ACTION, Method.CLICK, Object.TOOLBAR, Value.CAPTURE)
+                .param(Extra.VERSION, Integer.toString(TELEMETRY_VERSION))
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: Same here

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java (hunk 13 above)
Comment (mTwTm) [reply app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: Maybe rename TELEMETRY_VERSION to TOOL_BAR_CAPTURE_TELEMETRY_VERSION

File: app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java [hunk 15]
```diff
@@ -284,8 +292,29 @@ public static void settingsEvent(String key, String value) {
                 .queue();
     }
 
-    public static void settingsClickEvent(String key) {
+    public static void settingsEvent(Context context, String key, String value) {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/TelemetryWrapper.java]: I think this one is tricky. Write some comment to explain we only want pref_key_turbo_mode in firebase?

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+        VALID_PARAM_NAME = prepareString(FirebaseEvent.MAX_LENGTH_PARAM_NAME);
+
+        INVALID_PARAM_VALUE = prepareString(FirebaseEvent.MAX_LENGTH_PARAM_VALUE + 1);
+        VALID_PARAM_VALUE = prepareString(FirebaseEvent.MAX_LENGTH_PARAM_VALUE);
+
+    }
+
+    private static String prepareString(final int size) {
+        char[] chars = new char[size];
+        Arrays.fill(chars, '*');
```
Comment (mTwTm) [inline app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: Isn't '*' itself an invalid character?

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        return FirebaseEvent.create("", "", "", VALID_EVENT_NAME);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void eventNameLongerThan40Characters_shouldThrowIllegalArgumentException() {
+        FirebaseEvent.create("", "", "", INVALID_EVENT_NAME);
+    }
+
+    @Test
+    public void eventNameValid_shouldBeSafe() {
```
Comment (mTwTm) [inline app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: This check is doing nothing meaningful since FirebaseEvent.create() never returns null.

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+    public void nullContext_shouldBeSafe() {
+        generateValidFirebaseEvent().queue(null);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void paramSizeTooLarge_shouldThrow() {
+        final FirebaseEvent firebaseEvent = generateValidFirebaseEvent();
+        final Bundle extras = mock(Bundle.class);
+        firebaseEvent.setParam(extras);
+        when(extras.size()).thenReturn(FirebaseEvent.MAX_PARAM_SIZE + 1);
```
Comment (mTwTm) [inline app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: What is extras.size()? What is this testing against?
https://firebase.google.com/docs/reference/android/com/google/firebase/analytics/FirebaseAnalytics#logEvent(java.lang.String,%20android.os.Bundle)

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+
+    @Test
+    public void paramSizeValid_shouldBeSafe() {
+        final FirebaseEvent firebaseEvent = generateValidFirebaseEvent();
+        final Bundle extras = mock(Bundle.class);
+        firebaseEvent.setParam(extras);
+
+        when(extras.size()).thenReturn(FirebaseEvent.MAX_PARAM_SIZE);
+        final FirebaseEvent param = firebaseEvent.param("", "");
+        Assert.assertNotNull(param);
```
Comment (mTwTm) [inline app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: Again, I'm uncertain if this will ever be null.

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+    public void paramKeyLengthTooLarge_shouldThrow() {
+        final FirebaseEvent firebaseEvent = generateValidFirebaseEvent();
+        firebaseEvent.param(INVALID_PARAM_NAME, "");
+    }
+
+    @Test
+    public void paramKeyValid_shouldBeSafe() {
+        final FirebaseEvent firebaseEvent = generateValidFirebaseEvent();
+        final FirebaseEvent param = firebaseEvent.param(VALID_PARAM_NAME, "");
+        Assert.assertNotNull(param);
```
Comment (mTwTm) [inline app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: Same here.+

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+    public void paramValueLengthTooLarge_shouldThrow() {
+        final FirebaseEvent firebaseEvent = generateValidFirebaseEvent();
+        firebaseEvent.param("", INVALID_PARAM_VALUE);
+    }
+
+    @Test
+    public void paramValueLengthValid_shouldBeSafe() {
+        final FirebaseEvent firebaseEvent = generateValidFirebaseEvent();
+        final FirebaseEvent param = firebaseEvent.param("", VALID_PARAM_VALUE);
+        Assert.assertNotNull(param);
```
Comment (mTwTm) [inline app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: And here.

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java (hunk 16 above)
Comment (cnevinc) [reply app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: fixed

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java (hunk 18 above)
Comment (cnevinc) [reply app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: the number of the params should less than 25

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java (hunk 17 above)
Comment (cnevinc) [reply app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: Let me write it in another way.

File: app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java (hunk 11 above)
Comment (mTwTm) [reply app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java]: We will be notified for invalid telemetries? Sounds good enough to me. Just to make sure there is no hazard and no unexpected behaviors.

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java (hunk 18 above)
Comment (mTwTm) [reply app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: Oops my bad. I thought there's no bundle.size() but there is actually BaseBundle.size()

File: app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+    public static FirebaseEvent create(@NonNull String category, @NonNull String method, @Nullable String object) {
+        return new FirebaseEvent(category, method, object, null);
+    }
+
+    public FirebaseEvent param(String name, String value) {
+        if (this.eventParam == null) {
+            this.eventParam = new Bundle();
+        }
+        // only throws in non-release build
+        if (!AppConstants.isReleaseBuild() && this.eventParam.size() > MAX_PARAM_SIZE) {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java]: Should be >=, since we shouldn't add the 26th event if we already have 25.

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void paramKeyLengthTooLarge_shouldThrow() {
+        final FirebaseEvent firebaseEvent = generateValidFirebaseEvent();
+        firebaseEvent.param(INVALID_PARAM_NAME, "");
+    }
+
+    @Test
+    public void paramKeyValid_shouldBeSafe() {
```
Comment (mTwTm) [inline app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: I think this test is somewhat duplicating paramKeyLengthTooLarge_shouldThrow. Other pairs are duplications as well IMO.

Comment (cnevinc) [reply app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: fixed

File: app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java (hunk 22 above)
Comment (mTwTm) [reply app/src/main/java/org/mozilla/focus/telemetry/FirebaseEvent.java]: Maybe we should change the test case for this edge case too.

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java [hunk 24]
```diff
@@ -88,51 +88,18 @@ public void paramSizeTooLarge_shouldThrow() {
...
-
     @Test(expected = IllegalArgumentException.class)
     public void paramValueLengthTooLarge_shouldThrow() {
         final FirebaseEvent firebaseEvent = generateValidFirebaseEvent();
         firebaseEvent.param("", INVALID_PARAM_VALUE);
     }
 
-    @Test
-    public void paramValueLengthValid_shouldBeSafe() {
-        final FirebaseEvent firebaseEvent = generateValidFirebaseEvent();
```
Comment (mTwTm) [inline app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: I somewhat feel like keeping this is capturing more errors for us :P but it's up to you!

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+            fail("Should not have thrown any exception");
+        }
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void paramSizeTooLarge_shouldThrow() {
+        final FirebaseEvent firebaseEvent = generateValidFirebaseEvent();
+        final Bundle params = mock(Bundle.class);
+        firebaseEvent.setParam(params);
+        when(params.size()).thenReturn(FirebaseEvent.MAX_PARAM_SIZE + 1);
```
Comment (mTwTm) [inline app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: Let's also fix this to capture the edge case.

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java (hunk 24 above)
Comment (cnevinc) [reply app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: agree

File: app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java (hunk 25 above)
Comment (cnevinc) [reply app/src/testWebkit/java/org/mozilla/focus/telemetry/FirebaseEventTest.java]: fixed
